text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,315 @@
+    let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
+
+    for (_key, value) in record.iter_mut() {
+        let old_value = mem::take(value);
+        let analyzed = analyzer.analyze(&old_value);
+        
+        for (original, token) in analyzed.reconstruct() {
+            if token.is_word() {
+                let to_highlight = words.contains(&token.text());
+                if to_highlight { value.push_str("<mark>") }
+                value.push_str(original);
+                if to_highlight { value.push_str("</mark>") }
+            } else {
+                value.push_str(original);
+            }
+        }
+    }
+}
+```
+```rust
+// original
+fn highlight_record(record: &mut IndexMap<String, String>, words: &HashSet<String>) {
+    for (_key, value) in record.iter_mut() {
+        let old_value = mem::take(value);
+        for (token_type, token) in simple_tokenizer(&old_value) {
+            if token_type == TokenType::Word {
+                let lowercase_token = token.to_lowercase();
+                let to_highlight = words.contains(&lowercase_token);
+                if to_highlight { value.push_str("<mark>") }
+                value.push_str(token);
+                if to_highlight { value.push_str("</mark>") }
+            } else {
+                value.push_str(token);
+            }
+        }
+    }
+}
+```
+
+As we can see, the changes that need to be made are very minimal: this is because efforts have been made to make it's API close to the previous one.
+
+### Impact on documentation
+
+This feature should not impact the meilisearch user documentation,
+in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.
+
+## Technical Specifications
+
+### Architecture
+
+The new version of the tokenizer will replace the current version as a standalone library.
+
+### Implementation Details
+
+We want to support different tokenizers based on the language of the text to be indexed. For this we may need to change the tokenizer we are using while indexing, depending on the language and the script. The Analyzer provides an interface that abstract this need away from the consumer of the tokens.
+
+#### Pipeline
+
+A pipeline contains all the steps needed to tokenize a specific language or script,
+one or several preprocessors to prepare text before tokenization e.g. traditional chinese translation
+or erasing some characters;
+a tokeninizer to transform the text in an iterator of tokens;
+one or several normalizers to normalize tokens for indexation e.g. deunicode or lowercase.
+```rust
+pub struct Pipeline {
+    pre_processor: Box<dyn PreProcessor + 'static>,
+    tokenizer: Box<dyn Tokenizer + 'static>,
+    normalizer: Box<dyn Normalizer + 'static>,
+}
+```
+
+#### Configuration
+
+The tokenizer is configured at initialization. A `Default` implementation is provided with the settings we believe suits most cases.
```

Comment:
```suggestion
The tokenizer is configured at initialization. A `default` implementation is provided with the settings we believe suits most cases.
```

Reply:
default is the name of a trait, it start with an uppercase :)